File: AlkitabModel/src/yuku/alkitab/model/SingleChapterVerses.java
package yuku.alkitab.model;

public abstract class SingleChapterVerses {
	public static final String TAG = SingleChapterVerses.class.getSimpleName();
	
	private static final String[] VERSE_NUMBER_STRINGS = {
		"0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
		"10", "11", "12", "13", "14", "15", "16", "17", "18", "19",
		"20", "21", "22", "23", "24", "25", "26", "27", "28", "29",
		"30", "31", "32", "33", "34", "35", "36", "37", "38", "39",
		"40", "41", "42", "43", "44", "45", "46", "47", "48", "49",
		"50", "51", "52", "53", "54", "55", "56", "57", "58", "59",
		"60", "61", "62", "63", "64", "65", "66", "67", "68", "69",
		"70", "71", "72", "73", "74", "75", "76", "77", "78", "79",
	}; // up to [79]
	
	public abstract String getVerse(int verse_0);

	public abstract int getVerseCount();
	
	public String getVerseNumberText(int verse_0) {
		int verse_1 = verse_0 + 1;
		if (verse_1 < 80) {
			return VERSE_NUMBER_STRINGS[verse_1];
		} else {
			return Integer.toString(verse_1);
		}
	}
}

File: AlkitabModel/src/main/java/yuku/alkitab/model/SingleChapterVerses.kt
package yuku.alkitab.model

/**
 * Multiple verses that are presented as "single chapter".
 *
 * They are not necessarily in a real chapter, for example when presenting verses of different translation
 * at once, they are considered as if they were in a single chapter, only by making the [getVerseNumberText]
 * return the translation name instead of numbers.
 */
interface SingleChapterVerses {
    val verseCount: Int

    fun getVerse(verse_0: Int): String

    fun getVerseNumberText(verse_0: Int): String {
        val verse_1 = verse_0 + 1
        return if (verse_1 < 40) {
            VERSE_NUMBER_STRINGS[verse_1]
        } else {
            verse_1.toString()
        }
    }

    /**
     * Implement this when displaying verses in different font size for each verse.
     * For example, during comparison of different translations.
     */
    interface WithTextSizeMult : SingleChapterVerses {
        fun getTextSizeMult(verse_0: Int): Float
    }

    companion object {
        private val VERSE_NUMBER_STRINGS = arrayOf("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39") // up to [39]

        val EMPTY = object : SingleChapterVerses {
            override val verseCount: Int
                get() = 0

            override fun getVerse(verse_0: Int): String {
                throw IllegalArgumentException("@@getVerse requesting verse_0=$verse_0 but verseCount=$verseCount")
            }
        }
    }
}
